snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java

```diff
@@ -0,0 +1,204 @@
+
+  private Object conf;
+  private String catalogName = SnowflakeResources.DEFAULT_CATALOG_NAME;
+  private Map<String, String> catalogProperties = null;
+  private FileIO fileIO;
+  private QueryFactory queryFactory;
+
+  public SnowflakeCatalog() {}
+
+  public void setQueryFactory(QueryFactory factory) {
+    queryFactory = factory;
+  }
+
+  @Override
+  public List<TableIdentifier> listTables(Namespace namespace) {
+    Preconditions.checkArgument(
+        namespace.length() <= SnowflakeResources.MAX_NAMESPACE_DEPTH,
+        "Snowflake doesn't supports more than 2 levels of namespace");
+
+    List<SnowflakeTable> sfTables = queryFactory.listIcebergTables(namespace);
+
+    return sfTables.stream()
+        .map(
+            table ->
+                TableIdentifier.of(table.getDatabase(), table.getSchemaName(), table.getName()))
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public boolean dropTable(TableIdentifier identifier, boolean purge) {
+    return false;
+  }
+
+  @Override
+  public void renameTable(TableIdentifier from, TableIdentifier to) {}
+
+  @Override
+  public void initialize(String name, Map<String, String> properties) {
+    catalogProperties = properties;
+
+    String uri = properties.get(CatalogProperties.URI);
+    Preconditions.checkNotNull(uri, "JDBC connection URI is required");
+
+    if (name != null) {
+      this.catalogName = name;
+    }
+
+    LOG.debug("Connecting to JDBC database {}", properties.get(CatalogProperties.URI));
+
+    JdbcClientPool connectionPool = new JdbcClientPool(uri, properties);
+
+    if (queryFactory == null) {
+      queryFactory = new SnowflakeQueryFactory(connectionPool);
+    }
+
+    String fileIOImpl = SnowflakeResources.DEFAULT_FILE_IO_IMPL;
+
+    if (null != catalogProperties.get(CatalogProperties.FILE_IO_IMPL)) {
+      fileIOImpl = catalogProperties.get(CatalogProperties.FILE_IO_IMPL);
+    }
+
+    fileIO = CatalogUtil.loadFileIO(fileIOImpl, catalogProperties, conf);
+  }
+
+  @Override
+  public void close() {
+    queryFactory.close();
+  }
+
+  @Override
+  public void createNamespace(Namespace namespace, Map<String, String> metadata) {}
+
+  @Override
+  public List<Namespace> listNamespaces(Namespace namespace) {
+    Preconditions.checkArgument(
+        namespace.length() <= SnowflakeResources.MAX_NAMESPACE_DEPTH,
+        "Snowflake doesn't supports more than 2 levels of namespace");
+    List<Namespace> namespaceList = Lists.newArrayList();
+    try {
+      List<SnowflakeSchema> sfSchemas = queryFactory.listSchemas(namespace);
+      namespaceList =
+          sfSchemas.stream()
+              .map(schema -> Namespace.of(schema.getDatabase(), schema.getName()))
+              .collect(Collectors.toList());
+    } catch (UncheckedSQLException | UncheckedInterruptedException ex) {
+      LOG.error("{}", ex.getMessage(), ex);
+    }
+
+    if (namespace.length() == SnowflakeResources.MAX_NAMESPACE_DEPTH) {
```

Comment:
Looks like the logic here is actually for an edge case where we're trying to list schemas but already specified the exact schema we want. As in:

listNamespaces("foo_database.bar_schema");

and the logic in the JDBC layer currently just lists all schemas under "foo_database" and then this filter makes it so:

1. If the schema doesn't exist, we return empty list
2. If it does exist, we filter it down to a size-1 list "equal" to the input

Thinking about it more though, I'm not sure listing schemas in a schema should behave like an existence-check.

When listing namespaces under the "database" level, the 1-level database namespace isn't returned as one of the namespaces, even though the 1-level namespace is allowed in Spark's "in" clause:

    scala> spark.sql("show namespaces in iot_data").show(false);
    +---------------------------+
    |namespace                  |
    +---------------------------+
    |IOT_DATA.ICE_SCHEMA        |
    |IOT_DATA.INFORMATION_SCHEMA|
    |IOT_DATA.PUBLIC            |
    |IOT_DATA.TPCDS10TB         |
    +---------------------------+

So that implies to me that listing "in" a two-level namespace is supposed to only return 3-level namespaces, which aren't supported, and thus we should just throw IllegalArgumentException in that case.

Fixed the Precondition to require level <= MAX - 1